openTSNE/nearest_neighbors.py

```diff
@@ -445,3 +445,89 @@ def query(self, query, k):
 
         return indices, distances
 
+
+class HNSW(KNNIndex):
+    VALID_METRICS = [
+        "cosine",
+        "euclidean",
+        "dot",
+        "l2",
+        "ip",
+    ]
+
+    def __init__(self, *args, **kwargs):
+        try:
+            from hnswlib import Index  # pylint: disable=unused-import,unused-variable
+        except ImportError:
+            raise ImportError(
+                "Please install hnswlib: `conda install -c conda-forge "
+                "hnswlib` or `pip install hnswlib`."
+            )
+        super().__init__(*args, **kwargs)
+
+    def build(self, data, k: int):
+        timer = utils.Timer(
+            f"Finding {k} nearest neighbors using HNSWlib approximate search using "
+            f"{self.metric} distance...",
+            verbose=self.verbose,
+        )
+        timer.__enter__()
+
+        from hnswlib import Index
+
+        hnsw_space = {
+            "cosine": "cosine",
+            "dot": "ip",
+            "euclidean": "l2",
+            "ip": "ip",
+            "l2": "l2"
+        }[self.metric]
+
+        self.index = Index(space=hnsw_space, dim=data.shape[1])
+
+        metric_params = {
```

Comment:
@pavlin-policar Is this the intended use of `metric_params`? It does work, but the naming is a little bit confusing. Here, it is more used as `index_params`.

Reply:
No, you're using `metric_params` to pass parameters to the index construction. The idea of metric params is in case the metric we use takes additional params, as indicated by [`scikit-learn`](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors). I don't think I've ever had to use this.

We currently don't have any way to actually alter the index parameters, as you're doing here. Which kind of takes away from the flexibility I guess. But that kind of forces us to have good defaults.

Please change this though, since this is not the intedend usage of `metric_params.`

Reply:
Fair enough. I removed this part for now, it could stay in as an undocumented hack though :P
There seems to be no easy way to add the ability to configure the defaults from the outside, so I set the defaults from the HNSW examples.